snapcraft/commands/remote.py

```diff
@@ -407,12 +372,95 @@ def _monitor_and_complete(  # noqa: PLR0912, PLR0915
         )
         return return_code
 
-    def _get_project_build_fors(self) -> list[str]:
-        """Get a unique list of build-for architectures from the project.
+    def _get_archs(self, build_fors: list[str], platforms: list[str]) -> list[str]:
+        """Get the architectures to build for.
+
+        If the project contains a ``platforms`` or ``architectures`` key, then project's
+        build plan is used to determine the architectures to build for. The build plan
+        can be filtered using the ``--platform`` or ``--build-for`` arguments.
+
+        If the project doesn't contain a ``platforms`` or ``architectures`` key, then
+        the architectures to build for are defined by the ``--platform`` or
+        ``--build-for`` arguments.
+
+        If there are no architectures defined in the project or as arguments, then the
+        default behavior is to build for the host architecture.
+
+        :param build_fors: A list of build-for entries.
+        :param platforms: A list of platforms.
+
+        :raises EmptyBuildPlanError: If the build plan is filtered to an empty list.
+        :raises RemoteBuildError: If an unsupported architecture is provided.
 
         :returns: A list of architectures.
         """
-        build_fors = set({build_info.build_for for build_info in self.build_plan})
-        emit.debug(f"Parsed build-for architectures from build plan: {build_fors}")
+        archs: list[str] = []
+        # core22 projects with an `architectures` key will have a corresponding `platforms`
+        # key when the project is unmarshalled
+        if self.project.platforms or self.project._architectures_in_yaml:
+            # if the project has platforms, then the `--platforms` and `--build-for` arguments act as filters
+            if build_fors:
+                emit.debug("Filtering the build plan using the '--build-for' argument.")
+                for build_for in build_fors:
+                    filtered_build_plan = filter_plan(
+                        self.build_plan,
+                        platform=None,
+                        build_for=build_for,
+                        host_arch=None,
+                    )
+                    archs.extend([info.build_for for info in filtered_build_plan])
+                    if not archs:
+                        raise craft_application.errors.EmptyBuildPlanError()
+            elif platforms:
+                emit.debug("Filtering the build plan using the '--platforms' argument.")
+                for platform in platforms:
+                    filtered_build_plan = filter_plan(
+                        self.build_plan,
+                        platform=platform,
```

Comment:
it seems to me here that the build plan is filtered using the actual platform name, so how come it (the platform name) needs to be a deb architecture?

Reply:
Resolved via removing `--platform`.